### ISSUE TITLE
Introduce DependencyResolver

### ISSUE THREAD

**Issue by SamMousa** [yiisoft/di#56]
Currently dependency resolution is done in `AbstractContainer` and in `Injector`.

Instead we should create a `DependencyResolverInterface` (and implementation):
```php
interface DependencyResolverInterface {
    /**
     * @return Reference[]
     */
    public function resolveConstructor(string $class): array;

    /**
     * @return Reference[]
     */
    public function resolveCallable(callable $callable): array;

    /**
     * This is just an example, it doesn't add anything except syntax over resolveCallable.
     * @return Reference[]
     */
    public function resolveMethod($subject, string $method): array;

}

```
This approach will allow us to make the DI simpler to understand and simpler to test.

**Comment by samdark** [comment]
Agree.

**Comment by SamMousa** [comment]
Implemented in #57
